### PR TITLE
Reverse order of handshake for auth_method and encrypt_method

### PR DESCRIPTION
**Pull request by hirenvadalia**
#### Describe Bug or Feature

* Currently, if PBS_AUTH_METHOD and PBS_ENCRYPT_METHOD both are set then PBS will do handshake for PBS_AUTH_METHOD first then for PBS_ENCRYPT_METHOD, and data won't be encrypted till handshake of PBS_ENCRYPT_METHOD is completed.
This means if handshake data of PBS_AUTH_METHOD contains any sensitive information then it won't be encrypted.
But if we do handshake of PBS_ENCRYPT_METHOD first then we will have a secure channel while doing handshake of PBS_AUTH_METHOD, it means handshake data with sensitive information for PBS_AUTH_METHOD will be encrypted.
* (Found during testing) Currently TPP router code doesn't handle client-driven PBS_AUTH_METHOD and PBS_ENCRYPT_METHOD properly, it should use what client has provided in auth header, instead, it is using what is configured in local pbs.conf

#### Describe Your Change

* Changed code such a way that it will first do handshake for PBS_ENCRYPT_METHOD and establish a secure channel as soon as possible.
* Fixed TPP router for client-driven PBS_AUTH_METHOD and PBS_ENCRYPT_METHOD
* Modularized TPP handshake handling code

#### Link to Design Doc

* None, no change to external behavior

#### Attach Test and Valgrind Logs/Output

Screen Shot 2020-04-21 at 4 25 33 PM
Dummy libauth implementation: [dummy.txt](https://github.com/PBSPro/pbspro/files/4510796/dummy.txt)
[before.txt](https://github.com/PBSPro/pbspro/files/4510800/before.txt)
[after.txt](https://github.com/PBSPro/pbspro/files/4510798/after.txt)
(Notice `===== This is marker =====` in write() for fd 3)

### CODE REVIEW

**Comment by hirenvadalia** [comment]
Found some issue whilte testing, marking this PR as WIP

**Comment by hirenvadalia** [comment]
Test results based on the latest changes:
Screen Shot 2020-04-22 at 7 22 17 PM

**Comment by hirenvadalia** [comment]
Valgrind logs:
[after.txt](https://github.com/PBSPro/pbspro/files/4517991/after.txt)
[before.txt](https://github.com/PBSPro/pbspro/files/4517992/before.txt)

**Comment by hirenvadalia** [comment]
Manual Valgrind logs:
Setup: 1 host with server/comm/mom/sched, 1 host with only mom (attached to comm2), 1 host with only comm(comm2)
[python-valgrind.txt](https://github.com/PBSPro/pbspro/files/4522394/python-valgrind.txt) suppression file used to suppress Python-related leaks
Valgrind logs: [leaks.txt](https://github.com/PBSPro/pbspro/files/4522399/leaks.txt)

File: src/lib/Libtpp/tpp_internal.h [hunk 1]
```diff
@@ -413,8 +414,13 @@ typedef struct {
 	auth_def_t *authdef;
 	void *encryptctx;
 	auth_def_t *encryptdef;
+	pbs_auth_config_t *config;
+	int initiator;
```
Comment (subhasisb) [inline src/lib/Libtpp/tpp_internal.h:418]: call it conn_initiator? Jus to clarify that it is about connection initiation?

File: src/lib/Libtpp/tpp_router.c [hunk 2]
```diff
@@ -537,177 +537,43 @@ router_post_connect_handler(int tfd, void *data, void *c, void *extra)
...
-		if (authdef == NULL) {
-			tpp_log_func(LOG_CRIT, __func__, "Failed to find authdef in post connect handler");
-			return -1;
-		}
-
-		authdef->set_config((const pbs_auth_config_t *)(tpp_conf->auth_config));
-
-		if (authdef->create_ctx(&authctx, AUTH_CLIENT, AUTH_SERVICE_CONN, tpp_transport_get_conn_hostname(tfd))) {
-			tpp_log_func(LOG_CRIT, __func__, "Failed to create client auth context");
+		int conn_fd = ((tpp_router_t *) ctx->ptr)->conn_fd;
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply src/lib/Libtpp/tpp_router.c:551]: This is at top of the block in if condition.

[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_router.c:551]: ah ok ok yes true

**Pull request merged**